File: scripts/search/maryland_captcha_service_search.js
// Maryland DLLR License Search with CAPTCHA Solving Service
// This script uses a CAPTCHA solving service to automate searches
// Requires: API key from 2Captcha, Anti-Captcha, or CapSolver

const { chromium } = require('playwright');

// Configuration
const MARYLAND_URL = 'https://www.dllr.state.md.us/cgi-bin/ElectronicLicensing/OP_Search/OP_search.cgi?calling_app=RE::RE_personal_name';
const CAPTCHA_SERVICE_API_KEY = process.env.CAPTCHA_API_KEY || 'YOUR_API_KEY_HERE';
const CAPTCHA_SERVICE = process.env.CAPTCHA_SERVICE || '2captcha'; // Options: '2captcha', 'anticaptcha', 'capsolver'

// Employees to search
const EMPLOYEES = [
  { name: 'Edward Hyland', last_name: 'Hyland', first_name: 'Edward', priority: 'HIGH' },
  { name: 'Robert Kettler', last_name: 'Kettler', first_name: 'Robert', priority: 'HIGH' },
  { name: 'Caitlin Skidmore', last_name: 'Skidmore', first_name: 'Caitlin', priority: 'HIGH' },
  { name: 'Cindy Fisher', last_name: 'Fisher', first_name: 'Cindy', priority: 'MEDIUM' },
  { name: 'Luke Davis', last_name: 'Davis', first_name: 'Luke', priority: 'MEDIUM' },
  { name: 'Pat Cassada', last_name: 'Cassada', first_name: 'Pat', priority: 'MEDIUM' },
  { name: 'Sean Curtin', last_name: 'Curtin', first_name: 'Sean', priority: 'MEDIUM' },
  { name: 'Amy Groff', last_name: 'Groff', first_name: 'Amy', priority: 'MEDIUM' },
  { name: 'Robert Grealy', last_name: 'Grealy', first_name: 'Robert', priority: 'MEDIUM' },
  { name: 'Djene Moyer', last_name: 'Moyer', first_name: 'Djene', priority: 'MEDIUM' },
  { name: 'Henry Ramos', last_name: 'Ramos', first_name: 'Henry', priority: 'MEDIUM' },
  { name: 'Kristina Thoummarath', last_name: 'Thoummarath', first_name: 'Kristina', priority: 'MEDIUM' },
  { name: 'Christina Chang', last_name: 'Chang', first_name: 'Christina', priority: 'MEDIUM' },
  { name: 'Todd Bowen', last_name: 'Bowen', first_name: 'Todd', priority: 'MEDIUM' },
  { name: 'Jeffrey Williams', last_name: 'Williams', first_name: 'Jeffrey', priority: 'MEDIUM' }
];

// Solve CAPTCHA using service
async function solveCaptcha(page) {
  console.log(`Solving CAPTCHA using ${CAPTCHA_SERVICE}...`);

  try {
    // Method 1: Using playwright-recaptcha library (if installed)
    // const { solveRecaptcha } = require('playwright-recaptcha');
    // await solveRecaptcha(page, {
    //   apiKey: CAPTCHA_SERVICE_API_KEY,
    //   provider: CAPTCHA_SERVICE
    // });

    // Method 2: Direct API integration with 2Captcha
    if (CAPTCHA_SERVICE === '2captcha') {
      return await solveWith2Captcha(page);
    }

    // Method 3: Manual fallback - wait for user to complete
    console.log('CAPTCHA service not configured. Waiting for manual completion...');
    await page.waitForFunction(
      () => {
        const button = document.querySelector('button[type="submit"], input[type="submit"]');
        return button && !button.disabled;
      },
      { timeout: 300000 } // 5 minutes
    );
    return true;

  } catch (error) {
    console.error('CAPTCHA solving error:', error);
    return false;
  }
}

// 2Captcha integration
async function solveWith2Captcha(page) {
  const axios = require('axios');

  // Get site key from page
  const siteKey = await page.evaluate(() => {
    const recaptchaFrame = document.querySelector('iframe[src*="recaptcha"]');
    if (recaptchaFrame) {
      const src = recaptchaFrame.src;
      const match = src.match(/k=([^&]+)/);
      return match ? match[1] : null;
    }
    return null;
  });

  if (!siteKey) {
    throw new Error('Could not find reCAPTCHA site key');
  }

  // Get page URL
  const pageUrl = page.url();

  // Submit CAPTCHA to 2Captcha
  const submitResponse = await axios.post('http://2captcha.com/in.php', null, {
    params: {
      key: CAPTCHA_SERVICE_API_KEY,
      method: 'userrecaptcha',
      googlekey: siteKey,
      pageurl: pageUrl,
      json: 1
    }
  });

  if (submitResponse.data.status !== 1) {
    throw new Error(`2Captcha submit failed: ' + submitResponse.data.request`);
  }

  const taskId = submitResponse.data.request;
  console.log(`CAPTCHA task submitted. Task ID: ${taskId}`);

  // Poll for solution
  let solution = null;
  for (let i = 0; i < 60; i++) {
    await new Promise(resolve => setTimeout(resolve, 5000)); // Wait 5 seconds

    const resultResponse = await axios.get('http://2captcha.com/res.php', {
      params: {
        key: CAPTCHA_SERVICE_API_KEY,
        action: 'get',
        id: taskId,
        json: 1
      }
    });

    if (resultResponse.data.status === 1) {
      solution = resultResponse.data.request;
      break;
    } else if (resultResponse.data.request !== 'CAPCHA_NOT_READY') {
      throw new Error(`2Captcha error: ${resultResponse.data.request}`);
    }
  }

  if (!solution) {
    throw new Error('CAPTCHA solution timeout');
  }

  // Inject solution into page
  await page.evaluate((token) => {
    document.getElementById('g-recaptcha-response').innerHTML = token;
    const callback = window[`___grecaptcha_cfg`].clients[0].callback;
    if (callback) {
      callback(token);
    }
  }, solution);

  // Wait for button to become enabled
  await page.waitForFunction(
    () => {
      const button = document.querySelector('button[type="submit"], input[type="submit"]');
      return button && !button.disabled;
    },
    { timeout: 10000 }
  );

  return true;
}

// Search for employee
async function searchEmployee(page, employee) {
  console.log(`\nSearching for: ${employee.name}`);

  try {
    // Navigate to search page
    await page.goto(MARYLAND_URL);
    await page.waitForTimeout(2000);

    // Fill in Last Name
    const lastNameInput = page.locator('input[type="text"]').first();
    await lastNameInput.fill(employee.last_name);
    await page.waitForTimeout(1000);

    // Solve CAPTCHA
    const captchaSolved = await solveCaptcha(page);
    if (!captchaSolved) {
      throw new Error('CAPTCHA solving failed');
    }

    // Click Search button
    const searchButton = page.locator('button[type="submit"], input[type="submit"]');
    await searchButton.click();
    await page.waitForTimeout(3000);

    // Extract results
    const results = await page.evaluate(() => {
      const allText = document.body.innerText;
      const hasEmployee = allText.toLowerCase().includes(employee.last_name.toLowerCase());
      const resultCount = allText.match(/(\d+)\s+results? found/i);
      const hasNoResults = allText.toLowerCase().includes('no results') ||
                          allText.toLowerCase().includes('no records');

      // Try to extract license details from table
      const tables = Array.from(document.querySelectorAll('table'));
      let licenseData = null;

      for (const table of tables) {
        const rows = Array.from(table.querySelectorAll('tr'));
        if (rows.length > 1 && rows[0].innerText.toLowerCase().includes('license')) {
          licenseData = rows.slice(1).map(row => {
            const cells = Array.from(row.querySelectorAll('td, th'));
            return cells.map(cell => cell.innerText.trim());
          });
          break;
        }
      }

      return {
        found: hasEmployee,
        resultCount: resultCount ? parseInt(resultCount[1]) : (hasNoResults ? 0 : 'checking'),
        licenseData: licenseData,
        pageText: allText.substring(0, 2000)
      };
    });

    return {
      employee: employee.name,
      status: 'completed',
      results: results
    };

  } catch (error) {
    console.error(`Error searching for ${employee.name}:`, error);
    return {
      employee: employee.name,
      status: 'error',
      error: error.message
    };
  }
}

// Main execution
async function main() {
  console.log('=== Maryland DLLR License Search with CAPTCHA Service ===\n');
  console.log(`Service: ${CAPTCHA_SERVICE}`);
  console.log(`Total Employees: ${EMPLOYEES.length}\n`);

  if (CAPTCHA_SERVICE_API_KEY === 'YOUR_API_KEY_HERE') {
    console.log('⚠️  WARNING: CAPTCHA_API_KEY not set. Using manual CAPTCHA mode.');
    console.log('Set environment variable: export CAPTCHA_API_KEY=your_key_here\n');
  }

  const browser = await chromium.launch({ headless: false });
  const context = await browser.newContext();
  const page = await context.newPage();

  const results = [];

  // Search high priority first
  const highPriority = EMPLOYEES.filter(e => e.priority === 'HIGH');
  const mediumPriority = EMPLOYEES.filter(e => e.priority === 'MEDIUM');

  console.log('=== HIGH PRIORITY SEARCHES ===');
  for (const employee of highPriority) {
    const result = await searchEmployee(page, employee);
    results.push(result);
    await page.waitForTimeout(2000); // Rate limiting
  }

  console.log('\n=== MEDIUM PRIORITY SEARCHES ===');
  for (const employee of mediumPriority) {
    const result = await searchEmployee(page, employee);
    results.push(result);
    await page.waitForTimeout(2000); // Rate limiting
  }

  await browser.close();

  // Save results
  const fs = require('fs');
  const path = require('path');
  const outputDir = path.join(__dirname, '../../research/license_searches/maryland');
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  const outputFile = path.join(outputDir, 'maryland_captcha_service_results.json');
  fs.writeFileSync(outputFile, JSON.stringify({
    metadata: {
      date: new Date().toISOString(),
      service: CAPTCHA_SERVICE,
      total_searches: EMPLOYEES.length,
      completed: results.filter(r => r.status === 'completed').length,
      errors: results.filter(r => r.status === 'error').length
    },
    results: results
  }, null, 2));

  console.log(`\n✅ Results saved to: ${outputFile}`);
  console.log(`Completed: ${results.filter(r => r.status === 'completed').length}/${EMPLOYEES.length}`);
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = { searchEmployee, solveCaptcha };

File: scripts/search/maryland_captcha_service_search.R
#!/usr/bin/env Rscript
# Maryland DLLR License Search with CAPTCHA Solving Service
# This script provides framework for automated searches using CAPTCHA solving service

library(jsonlite)

# Configuration
current_dir <- getwd()
while (!file.exists(file.path(current_dir, "README.md")) && current_dir != "/") {
  current_dir <- dirname(current_dir)
}
PROJECT_ROOT <- if (file.exists(file.path(current_dir, "README.md"))) current_dir else getwd()

RESEARCH_DIR <- file.path(PROJECT_ROOT, "research", "license_searches", "maryland")
dir.create(RESEARCH_DIR, showWarnings = FALSE, recursive = TRUE)

# CAPTCHA Service Configuration
CAPTCHA_SERVICES <- list(
  twocaptcha = list(
    name = "2Captcha",
    url = "https://2captcha.com",
    api_endpoint = "http://2captcha.com/in.php",
    result_endpoint = "http://2captcha.com/res.php",
    cost_per_1000 = 2.99,
    cost_per_search = 0.00299
  ),
  anticaptcha = list(
    name = "Anti-Captcha",
    url = "https://anti-captcha.com",
    api_endpoint = "https://api.anti-captcha.com/createTask",
    result_endpoint = "https://api.anti-captcha.com/getTaskResult",
    cost_per_1000 = 1.00,
    cost_per_search = 0.001
  ),
  capsolver = list(
    name = "CapSolver",
    url = "https://capsolver.com",
    api_endpoint = "https://api.capsolver.com/createTask",
    result_endpoint = "https://api.capsolver.com/getTaskResult",
    cost_per_1000 = 1.50,
    cost_per_search = 0.0015
  )
)

# Employees to search (MEDIUM priority - after high priority manual searches)
MEDIUM_PRIORITY_EMPLOYEES <- list(
  list(name = "Cindy Fisher", last_name = "Fisher", first_name = "Cindy"),
  list(name = "Luke Davis", last_name = "Davis", first_name = "Luke"),
  list(name = "Pat Cassada", last_name = "Cassada", first_name = "Pat"),
  list(name = "Sean Curtin", last_name = "Curtin", first_name = "Sean"),
  list(name = "Amy Groff", last_name = "Groff", first_name = "Amy"),
  list(name = "Robert Grealy", last_name = "Grealy", first_name = "Robert"),
  list(name = "Djene Moyer", last_name = "Moyer", first_name = "Djene"),
  list(name = "Henry Ramos", last_name = "Ramos", first_name = "Henry"),
  list(name = "Kristina Thoummarath", last_name = "Thoummarath", first_name = "Kristina"),
  list(name = "Christina Chang", last_name = "Chang", first_name = "Christina"),
  list(name = "Todd Bowen", last_name = "Bowen", first_name = "Todd"),
  list(name = "Jeffrey Williams", last_name = "Williams", first_name = "Jeffrey")
)

# Maryland DLLR reCAPTCHA Site Key
MARYLAND_RECAPTCHA_SITE_KEY <- "6LeUU6ApAAAAANxcBOW8c_zqA8mPt2fjzW1KY7aw"
MARYLAND_URL <- "https://www.dllr.state.md.us/cgi-bin/ElectronicLicensing/OP_Search/OP_search.cgi?calling_app=RE::RE_personal_name"

# Create service integration guide
create_service_guide <- function() {
  guide <- list(
    metadata = list(
      date = as.character(Sys.Date()),
      purpose = "CAPTCHA Solving Service Integration for Maryland DLLR",
      total_employees = length(MEDIUM_PRIORITY_EMPLOYEES),
      estimated_cost = list(
        twocaptcha = length(MEDIUM_PRIORITY_EMPLOYEES) * CAPTCHA_SERVICES$twocaptcha$cost_per_search,
        anticaptcha = length(MEDIUM_PRIORITY_EMPLOYEES) * CAPTCHA_SERVICES$anticaptcha$cost_per_search,
        capsolver = length(MEDIUM_PRIORITY_EMPLOYEES) * CAPTCHA_SERVICES$capsolver$cost_per_search
      )
    ),
    services = CAPTCHA_SERVICES,
    implementation = list(
      javascript_playwright = list(
        library = "playwright-recaptcha",
        install = "npm install playwright-recaptcha",
        example = paste0(
          "const { solveRecaptcha } = require('playwright-recaptcha');\n",
          "await solveRecaptcha(page, {\n",
          "  apiKey: 'YOUR_API_KEY',\n",
          "  provider: '2captcha' // or 'anticaptcha', 'capsolver'\n",
          "});"
        )
      ),
      python_playwright = list(
        library = "playwright-recaptcha",
        install = "pip install playwright-recaptcha",
        example = paste0(
          "from playwright_recaptcha import solve_recaptcha\n",
          "await solve_recaptcha(page, api_key='YOUR_API_KEY', provider='2captcha')"
        )
      )
    ),
    employees = MEDIUM_PRIORITY_EMPLOYEES,
    site_key = MARYLAND_RECAPTCHA_SITE_KEY,
    search_url = MARYLAND_URL
  )

  return(guide)
}

# Save guide
guide <- create_service_guide()
guide_file <- file.path(RESEARCH_DIR, "captcha_service_integration_guide.json")
write_json(guide, guide_file, pretty = TRUE, auto_unbox = TRUE)

cat("=== CAPTCHA Service Integration Guide Created ===\n")
cat("Date:", Sys.Date(), "\n")
cat("Total Employees (Medium Priority):", length(MEDIUM_PRIORITY_EMPLOYEES), "\n")
cat("Guide saved to:", guide_file, "\n\n")
cat("SERVICE OPTIONS:\n")
cat("1. 2Captcha - $", round(guide$metadata$estimated_cost$twocaptcha, 4), "for", length(MEDIUM_PRIORITY_EMPLOYEES), "searches\n")
cat("2. Anti-Captcha - $", round(guide$metadata$estimated_cost$anticaptcha, 4), "for", length(MEDIUM_PRIORITY_EMPLOYEES), "searches\n")
cat("3. CapSolver - $", round(guide$metadata$estimated_cost$capsolver, 4), "for", length(MEDIUM_PRIORITY_EMPLOYEES), "searches\n\n")
cat("NEXT STEPS:\n")
cat("1. Choose a CAPTCHA solving service\n")
cat("2. Sign up and get API key\n")
cat("3. Install playwright-recaptcha library\n")
cat("4. Run automated searches for medium priority employees\n")
